pi-hole/docker-pi-hole
# Allow running pihole-FTL as non-root

## Description
- Add a new container environment variable allowing to specify the user to run the pihole-FTL process as. Defaults to root.
- Set inherited capabilities attributes on the pihole-FTL file to automatically grant runtime permitted capabilities when available in the bounding set. This allows dropping root before starting pihole-FTL without failing with a permission error if the capabilities are not available to the container (the process may still error out if performing an operation requiring the capability).
- Add some information on capabilities to the Readme file.

## Motivation and Context


Running FTLDNS as non-root is recommended for security reasons.
It also causes dnsmasq to skip the capabilities syscalls, allowing the container to run with more limited capabilities (i.e. ability to not add NET_ADMIN, or even drop NET_RAW).
Fixes #414, #392.

## How Has This Been Tested?



Ran the automated tests but my development environment seem to have issues running them even before changes.
To test the changes I performed the following manual tests:
- set `DNSMASQ_USER=root`, use `cap-add=NET_ADMIN`. Verify process is still run as root with all capabilities from the bounding set as effective.
- set `DNSMASQ_USER=pihole`, use `cap-add=NET_ADMIN`. Verify process is run as `pihole` with the NET_BIND_SERVICE, NET_RAW and NET_ADMIN capabilities as effective. Verify that the container starts successfully and that FTLDNS answers queries.
- set `DNSMASQ_USER=pihole`, use `cap-drop=NET_RAW`. Verify process is run as `pihole` with only the NET_BIND_SERVICE capability as effective. Verify that the container starts successfully and that FTLDNS answers queries.
- set `DNSMASQ_USER=root`, use `cap-drop=NET_SETFCAP`. Verify that container starts and process runs as root. This is to simulate a failure in setting the file capabilities attributes, which doesn't matter if the process is run as root.
- set `DNSMASQ_USER=pihole`, use `cap-drop=NET_SETFCAP`. Verify container fails checks and outputs an error message explaining the container cannot run as non-root.

I guess those could be automated, but I'm really not familiar enough with python for that.

## Types of changes


## Checklist:
